src/Microsoft.ML.Featurizers/DateTimeTransformer.cs

```diff
@@ -0,0 +1,834 @@
+
+            private protected override bool CreateTransformerFromEstimatorHelper(TransformerEstimatorSafeHandle estimator, out IntPtr transformer, out IntPtr errorHandle) =>
+                CreateTransformerFromEstimatorNative(estimator, out transformer, out errorHandle);
+
+            private protected override bool DestroyEstimatorHelper(IntPtr estimator, out IntPtr errorHandle) =>
+                DestroyEstimatorNative(estimator, out errorHandle);
+
+            private protected override bool DestroyTransformerHelper(IntPtr transformer, out IntPtr errorHandle) =>
+                DestroyTransformerNative(transformer, out errorHandle);
+
+            [DllImport("Featurizers", EntryPoint = "DateTimeFeaturizer_CreateTransformerSaveData", CallingConvention = CallingConvention.Cdecl), SuppressUnmanagedCodeSecurity]
+            private static extern bool CreateTransformerSaveDataNative(TransformerEstimatorSafeHandle transformer, out IntPtr buffer, out IntPtr bufferSize, out IntPtr error);
+            private protected override bool CreateTransformerSaveDataHelper(out IntPtr buffer, out IntPtr bufferSize, out IntPtr errorHandle) =>
+                CreateTransformerSaveDataNative(_transformerHandler, out buffer, out bufferSize, out errorHandle);
+        }
+
+        #endregion
+
+        private sealed class Mapper : MapperBase
+        {
+
+            #region Class data members
+
+            private readonly DateTimeTransformer _parent;
+            private ConcurrentDictionary<long, TimePoint> _cache;
+            private ConcurrentQueue<long> _oldestKeys;
+
+            #endregion
+
+            public Mapper(DateTimeTransformer parent, DataViewSchema inputSchema) :
+                base(parent.Host.Register(nameof(Mapper)), inputSchema, parent)
+            {
+                _parent = parent;
+                _cache = new ConcurrentDictionary<long, TimePoint>();
+                _oldestKeys = new ConcurrentQueue<long>();
+            }
+
+            protected override DataViewSchema.DetachedColumn[] GetOutputColumnsCore()
+            {
+                var columns = new List<DataViewSchema.DetachedColumn>();
+
+                foreach (DateTimeEstimator.ColumnsProduced column in Enum.GetValues(typeof(DateTimeEstimator.ColumnsProduced)))
+                    if (_parent._columnsToDrop == null || !_parent._columnsToDrop.Contains(column))
+                    {
+                        columns.Add(new DataViewSchema.DetachedColumn(_parent._column.Prefix + column.ToString(),
+                            ColumnTypeExtensions.PrimitiveTypeFromType(column.GetRawColumnType())));
+                    }
+
+                return columns.ToArray();
+            }
+
+            private Delegate MakeGetter<T>(DataViewRow input, int iinfo)
+            {
+                ValueGetter<T> result = (ref T dst) =>
+                {
+                    long dateTime = default;
```

Comment:
Why is the input `dateTime` column a `long` and not a `DateTime`?

Reply:
`long` to match a unix timestamp. We will be adding in the ability to use C# `DateTime` and `string` in an ISO format in the future, but the initial implementation is just the unix timestamp.

Reply:
My concern here is that .NET developers expect to have their date time data using `System.DateTime`. That is the natural type in .NET. Forcing the user to convert from `System.DateTime` to a long unix timestamp doesn't seem like a good experience. It would make more sense if the input column was a `System.DateTime`.

Reply:
Why not do the conversion yourself here? The user's input column can be `DateTime`, and then you convert it to a long unix timestamp before passing it to the C++ code?

Reply:
We want to ensure that the process of converting from a string to "Date time" is consistent across frameworks - especially considering that a DateTime in and of itself isn't sufficient in some scenarios, as we lose (potentially important) time zone information in the process.

Reply:
I'm not asking for string. I'm asking for the ML.NET data type that a user would feed into this transform to be the canonical `System.DateTime` type and not a `long` type.

Specifically, this transform should require the input column to be of this type:

https://github.com/dotnet/machinelearning/blob/9fc8f1cd323d572672a5460914f927acfef310dc/src/Microsoft.ML.DataView/DataViewType.cs#L361-L365

It doesn't make sense for a user to have to convert a `DateTime` to a `long` before calling this transform.

Reply:
Specifically, look at the sample code being checked in under `samples`:

```C#
            // Create a small dataset as an IEnumerable.
            // Future Date - 2025 June 30
            var samples = new[] { new DateTimeInput() { Date = 1751241600 } };
```

Instead, that sample code should be:

```C#
            // Create a small dataset as an IEnumerable.
            var samples = new[] { new DateTimeInput() { Date = new DateTime(2025, 6, 30) } };
```

Reply:
Agreed - eventually a C# developer will be able to pass `System:DateTime`, `string`, or `long`.

Reply:
Strongly agree on using System.DateTime from the beginning. This will make the transform consistent with other parts of the API, as well as more usable to .NET developers

Reply:
That is on our todo list and will be added in in the future. We will eventually be supporting `DateTime` and `string` as well, but initial support will just be for `long`.